### ISSUE TITLE
Training with different image size

### ISSUE THREAD

**Issue by acaelles97** [facebookresearch/maskrcnn-benchmark#827]
## ❓ Questions and Help
Hi all,
I have changed the input image size (2560x2048) of config file to train with high quality masks, as i want them to be as good as possible. The training seemed fine but when testing on same image size  i got really bad detections (few of expected and confidence threshold really low) output following inference tutorial. Should i have changed more parameters on config file apart from INPUT: MAX_SIZE_TRAIN: 2560 MIN_SIZE_TRAIN: 2048   MAX_SIZE_TEST: 2560  MIN_SIZE_TEST: 2048?  Here's my training config. Test config is the same but changing WEIGHTS for model i want to test and adding TEST's size. 
Thanks a lot for you're help!!

2019-05-24 14:24:05,621 maskrcnn_benchmark INFO: Using 2 GPUs
2019-05-24 14:24:05,621 maskrcnn_benchmark INFO: Namespace(config_file='/shared/configs/e2e_ms_rcnn_R_50_FPN_1x_from_scratch_HQ_images.yaml', distributed=True, local_rank=0, opts=[], skip_test=False)
2019-05-24 14:24:05,621 maskrcnn_benchmark INFO: Collecting env info (might take some time)
2019-05-24 14:24:07,236 maskrcnn_benchmark INFO: 
PyTorch version: 1.1.0.dev20190516
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 9.0.176
GPU models and configuration: 
GPU 0: GeForce GTX 1080
GPU 1: GeForce GTX 1080

Nvidia driver version: 384.130
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.5.0

Versions of relevant libraries:
[pip] numpy==1.16.3
[pip] torch==1.1.0.dev20190516
[pip] torchvision==0.2.1
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.3                      199  
[conda] mkl_fft                   1.0.12           py37ha843d7b_0  
[conda] mkl_random                1.0.2            py37hd81dba3_0  
[conda] pytorch-nightly           1.1.0.dev20190516 py3.7_cuda9.0.176_cudnn7.5.1_0    pytorch
[conda] torchvision               0.2.1                    pypi_0    pypi
        Pillow (6.0.0)
2019-05-24 14:24:07,236 maskrcnn_benchmark INFO: Loaded configuration file /shared/configs/e2e_ms_rcnn_R_50_FPN_1x_from_scratch_HQ_images.yaml
2019-05-24 14:24:07,236 maskrcnn_benchmark INFO: 
MODEL:
  META_ARCHITECTURE: "GeneralizedRCNN"
  WEIGHT: https://dl.fbaipublicfiles.com/detectron/ImageNetPretrained/MSRA/R-50.pkl
  
  #~ WEIGHT: "/shared/Pre-weights/R-50.pkl"
  PRETRAINED_MODELS: 'pretrained_models'
  BACKBONE:
    CONV_BODY: "R-50-FPN"
    OUT_CHANNELS: 256
  RPN:
    USE_FPN: True
    ANCHOR_STRIDE: (4, 8, 16, 32, 64)
    PRE_NMS_TOP_N_TRAIN: 2000
    PRE_NMS_TOP_N_TEST: 1000
    POST_NMS_TOP_N_TEST: 1000
    FPN_POST_NMS_TOP_N_TRAIN: 1000
    FPN_POST_NMS_TOP_N_TEST: 1000
  ROI_HEADS:
    USE_FPN: True
  ROI_BOX_HEAD:
    POOLER_RESOLUTION: 7
    POOLER_SCALES: (0.25, 0.125, 0.0625, 0.03125)
    POOLER_SAMPLING_RATIO: 2
    FEATURE_EXTRACTOR: "FPN2MLPFeatureExtractor"
    PREDICTOR: "FPNPredictor"
    NUM_CLASSES: 3 
  
  ROI_MASK_HEAD:
    POOLER_SCALES: (0.25, 0.125, 0.0625, 0.03125)
    FEATURE_EXTRACTOR: "MaskRCNNFPNFeatureExtractor"
    PREDICTOR: "MaskRCNNC4Predictor"
    POOLER_RESOLUTION: 14
    POOLER_SAMPLING_RATIO: 2
    RESOLUTION: 28
    SHARE_BOX_FEATURE_EXTRACTOR: False
  MASK_ON: True
  MASKIOU_ON: True
DATASETS:
  TRAIN: ("D42-20_HR", "D42-19_HR", "D42-18_HR", "D42-17_HR", "D42-16_HR",)  
  TEST: ("D42-02-val",)
DATALOADER:
  SIZE_DIVISIBILITY: 32
SOLVER:
  BASE_LR: 0.005
  IMS_PER_BATCH: 2 
  WEIGHT_DECAY: 0.0001
  #~ STEPS: (60000, 80000)
  STEPS: (10000, 20000)
  #~ MAX_ITER: 90000
  MAX_ITER: 29000 #20000 1 epoch
  CHECKPOINT_PERIOD: 2500
TEST:
  IMS_PER_BATCH: 1

INPUT:
  MAX_SIZE_TRAIN: 2560
  MIN_SIZE_TRAIN: 2048

OUTPUT_DIR: "/shared/Training/T-13_HQ"

2019-05-24 14:24:07,237 maskrcnn_benchmark INFO: Running with config:
DATALOADER:
  ASPECT_RATIO_GROUPING: True
  NUM_WORKERS: 4
  SIZE_DIVISIBILITY: 32
DATASETS:
  TEST: ('D42-02-val',)
  TRAIN: ('D42-20_HR', 'D42-19_HR', 'D42-18_HR', 'D42-17_HR', 'D42-16_HR')
INPUT:
  MAX_SIZE_TEST: 1333
  MAX_SIZE_TRAIN: 2560
  MIN_SIZE_TEST: 800
  MIN_SIZE_TRAIN: 2048
  PIXEL_MEAN: [102.9801, 115.9465, 122.7717]
  PIXEL_STD: [1.0, 1.0, 1.0]
  TO_BGR255: True
MODEL:
  BACKBONE:
    CONV_BODY: R-50-FPN
    FREEZE_CONV_BODY_AT: 2
    OUT_CHANNELS: 256
  DEVICE: cuda
  MASKIOU_LOSS_WEIGHT: 1.0
  MASKIOU_ON: True
  MASK_ON: True
  META_ARCHITECTURE: GeneralizedRCNN
  PRETRAINED_MODELS: pretrained_models
  RESNETS:
    NUM_GROUPS: 1
    RES2_OUT_CHANNELS: 256
    RES5_DILATION: 1
    STEM_FUNC: StemWithFixedBatchNorm
    STEM_OUT_CHANNELS: 64
    STRIDE_IN_1X1: True
    TRANS_FUNC: BottleneckWithFixedBatchNorm
    WIDTH_PER_GROUP: 64
  ROI_BOX_HEAD:
    FEATURE_EXTRACTOR: FPN2MLPFeatureExtractor
    MLP_HEAD_DIM: 1024
    NUM_CLASSES: 3
    POOLER_RESOLUTION: 7
    POOLER_SAMPLING_RATIO: 2
    POOLER_SCALES: (0.25, 0.125, 0.0625, 0.03125)
    PREDICTOR: FPNPredictor
  ROI_HEADS:
    BATCH_SIZE_PER_IMAGE: 512
    BBOX_REG_WEIGHTS: (10.0, 10.0, 5.0, 5.0)
    BG_IOU_THRESHOLD: 0.5
    DETECTIONS_PER_IMG: 100
    FG_IOU_THRESHOLD: 0.5
    NMS: 0.5
    POSITIVE_FRACTION: 0.25
    SCORE_THRESH: 0.05
    USE_FPN: True
  ROI_MASK_HEAD:
    CONV_LAYERS: (256, 256, 256, 256)
    FEATURE_EXTRACTOR: MaskRCNNFPNFeatureExtractor
    MLP_HEAD_DIM: 1024
    POOLER_RESOLUTION: 14
    POOLER_SAMPLING_RATIO: 2
    POOLER_SCALES: (0.25, 0.125, 0.0625, 0.03125)
    POSTPROCESS_MASKS: False
    POSTPROCESS_MASKS_THRESHOLD: 0.5
    PREDICTOR: MaskRCNNC4Predictor
    RESOLUTION: 28
    SHARE_BOX_FEATURE_EXTRACTOR: False
  RPN:
    ANCHOR_SIZES: (32, 64, 128, 256, 512)
    ANCHOR_STRIDE: (4, 8, 16, 32, 64)
    ASPECT_RATIOS: (0.5, 1.0, 2.0)
    BATCH_SIZE_PER_IMAGE: 256
    BG_IOU_THRESHOLD: 0.3
    FG_IOU_THRESHOLD: 0.7
    FPN_POST_NMS_TOP_N_TEST: 1000
    FPN_POST_NMS_TOP_N_TRAIN: 1000
    MIN_SIZE: 0
    NMS_THRESH: 0.7
    POSITIVE_FRACTION: 0.5
    POST_NMS_TOP_N_TEST: 1000
    POST_NMS_TOP_N_TRAIN: 2000
    PRE_NMS_TOP_N_TEST: 1000
    PRE_NMS_TOP_N_TRAIN: 2000
    RPN_HEAD: SingleConvRPNHead
    STRADDLE_THRESH: 0
    USE_FPN: True
  RPN_ONLY: False
  WEIGHT: https://dl.fbaipublicfiles.com/detectron/ImageNetPretrained/MSRA/R-50.pkl
OUTPUT_DIR: /shared/Training/T-13_HQ
PATHS_CATALOG: /home/maskscoring_rcnn/maskrcnn_benchmark/config/paths_catalog.py
SOLVER:
  BASE_LR: 0.005
  BIAS_LR_FACTOR: 2
  CHECKPOINT_PERIOD: 2500
  GAMMA: 0.1
  IMS_PER_BATCH: 2
  MAX_ITER: 29000
  MOMENTUM: 0.9
  STEPS: (10000, 20000)
  WARMUP_FACTOR: 0.3333333333333333
  WARMUP_ITERS: 500
  WARMUP_METHOD: linear
  WEIGHT_DECAY: 0.0001
  WEIGHT_DECAY_BIAS: 0
TEST:
  EXPECTED_RESULTS: []
  EXPECTED_RESULTS_SIGMA_TOL: 4
  IMS_PER_BATCH: 1

**Comment by suica** [comment]
If I don't misunderstand your description ("following inference tutorial"), you used code in `maskrcnn-benchmark/demo/predictor.py`.

In old versions of this code, users should manually pass min_image_size to COCODemo() instead of just using MAX/MIN_SIZE_TEST in specified config to resize the input. So there could be the problem.

I will consider a recheck to whether you run inference with the latest copy of both `predictor.py` :)

**Comment by acaelles97** [comment]
Thanks for you're comment!
When creating the  COCODemo() instance, i'm actually passing the desired min image size, so i guess the problem might be somewhere else.

coco_demo = COCODemo(
    cfg,
    min_image_size=2048,
    confidence_threshold=0.7,
)

**Comment by suica** [comment]
How about the version of `predictor.py`? It does matter.

And, as another way to locate the problem, I recommend you to use `print(image.shape)`  after line 230 of `predictor.py` (https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/demo/predictor.py#L230),  then check the shape to make sure the input of network is resized as expected.

**Comment by acaelles97** [comment]
I have just update my predictor.py version, so now should be up-to-date. I have also checked the image shape after transform is applied, and it is as expected:
torch.Size([3, 2048, 2560])
Still does not work as expected at all, don't know what might be wrong

**Comment by fmassa** [comment]
You config has the following arguments:
```
MAX_SIZE_TEST: 1333
MAX_SIZE_TRAIN: 2560
MIN_SIZE_TEST: 800
MIN_SIZE_TRAIN: 2048
```
Check that you are indeed modifying the test size.

**Comment by acaelles97** [comment]
Sorry for not making it clear.
I am using another config file for inference, which is basically the same changing WEIGHTS for the ones i want to do inference with and defining correctly MAX_SIZE_TEST and MIN_SIZE_TEST. 
Am I missing something else?
Thanks for you're repply!

MODEL:
  META_ARCHITECTURE: "GeneralizedRCNN"
  WEIGHT: /shared/Training/T-13_HQ/model_0020000.pth
  PRETRAINED_MODELS: 'pretrained_models'
  BACKBONE:
    CONV_BODY: "R-50-FPN"
    OUT_CHANNELS: 256
  RPN:
    USE_FPN: True
    ANCHOR_STRIDE: (4, 8, 16, 32, 64)
    PRE_NMS_TOP_N_TRAIN: 2000
    PRE_NMS_TOP_N_TEST: 1000
    POST_NMS_TOP_N_TEST: 1000
    FPN_POST_NMS_TOP_N_TRAIN: 1000
    FPN_POST_NMS_TOP_N_TEST: 1000
  ROI_HEADS:
    USE_FPN: True
  ROI_BOX_HEAD:
    POOLER_RESOLUTION: 7
    POOLER_SCALES: (0.25, 0.125, 0.0625, 0.03125)
    POOLER_SAMPLING_RATIO: 2
    FEATURE_EXTRACTOR: "FPN2MLPFeatureExtractor"
    PREDICTOR: "FPNPredictor"
    NUM_CLASSES: 3 
  
  ROI_MASK_HEAD:
    POOLER_SCALES: (0.25, 0.125, 0.0625, 0.03125)
    FEATURE_EXTRACTOR: "MaskRCNNFPNFeatureExtractor"
    PREDICTOR: "MaskRCNNC4Predictor"
    POOLER_RESOLUTION: 14
    POOLER_SAMPLING_RATIO: 2
    RESOLUTION: 28
    SHARE_BOX_FEATURE_EXTRACTOR: False
  MASK_ON: True
  MASKIOU_ON: True
DATASETS:
  TRAIN: ("D42-20_HR", "D42-19_HR", "D42-18_HR", "D42-17_HR", "D42-16_HR",)  
  TEST: ("D42-02-val",)
DATALOADER:
  SIZE_DIVISIBILITY: 32
SOLVER:
  BASE_LR: 0.005
  IMS_PER_BATCH: 2 
  WEIGHT_DECAY: 0.0001
  #~ STEPS: (60000, 80000)
  STEPS: (10000, 20000)
  #~ MAX_ITER: 90000
  MAX_ITER: 29000 #20000 1 epoch
  CHECKPOINT_PERIOD: 2500
TEST:
  IMS_PER_BATCH: 1

INPUT:
  MAX_SIZE_TRAIN: 2560
  MIN_SIZE_TRAIN: 2048
  MAX_SIZE_TEST: 2560
  MIN_SIZE_TEST: 2048


OUTPUT_DIR: "/shared/Training/T-13_HQ"

**Comment by tangze5258** [comment]
hi bro,do u konw how to modify these parameters MIN_SIZE_TRAIN, MAX_SIZE_TRAIN, if my custom imgs shape larger than 1333*800?

**Comment by acaelles97** [comment]
Shortest edge of your image will be rescaled to the MIN_SIZE_TRAIN value you put on your config file. The largest edge will be rescaled according to the aspect ratio of your image up to the MAX_SIZE_TRAIN value. Unless your task is very dependent on the resolution of your images( you need to work with hig resolution images), just don't modify anything and your images will be automatically rescaled to the default values. 
Then on test time just be sure your image test is same as train as expained above, and should work correctly (I dont have any problem working on default image size).

**Comment by tangze5258** [comment]
@acaelles97 @fmassa 
thank u, i'll try it again.

**Comment by acaelles97** [comment]
I have just discovered that training masks were correct but bounding box were not. I fixed that and now working really well.
Thanks for your comments.

**Comment by pangsu0613** [comment]
Hello @acaelles97 , could you tell me how you fix the uncorrected bounding box issue? Thank you very much.
